### PR TITLE
Refactor/implement collection iterator & add benchmark

### PR DESCRIPTION
**Pull request by hannunosto**
* Introduce iterator to perform actions on collections / collection items
* Split Index service into two separate services - Index service takes care of invalidating and rebuilding product data, Sync service handles sending the product data to Nosto 
* Add performance benchmarking for data processing

## Motivation and Context
Make the code more readable and easier to test.

## How Has This Been Tested?
Tested locally by running the indexers.

## Documentation:
To be done later on. 

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] I have assigned the correct milestone or created one if non-existent.
- [x] I have correctly labeled this pull request.
- [x] I have linked the corresponding issue in this description.
- [x] I have updated the corresponding Jira ticket.
- [x] I have requested a review from at least 2 reviewers
- [x] I have checked the base branch of this pull request
- [x] I have checked my code for any possible security vulnerabilities

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ All notable changes to this project will be documented in this file. This projec
 * Use graphql for sending order confirmations and order status updates
 * Speedup the SKU price lookups by using price index table (catalog_product_index_price)
 
+### 3.8.5
```
Comment (supercid) [inline CHANGELOG.md]: Bump it up, we already release `3.8.6`

File: Model/Service/Sync.php [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                $op->upsert();
+                $this->tickBenchmark(self::BENCHMARK_SYNC_NAME);
+            } catch (\Exception $upsertException) {
+                $this->getLogger()->exception($upsertException);
+            } finally {
+                // We will set this as in sync even if there was failures
+                $collectionBatch->markAsInSyncCurrentItemsByStore($store);
+            }
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_SYNC_NAME, $store);
```
Comment (supercid) [inline Model/Service/Sync.php]: Try-catch?

File: Model/Service/Sync.php [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /**
+     * Discontinues products in Nosto and removes indexed products from Nosto product index
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @return int number of deleted products
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
```
Comment (supercid) [inline Model/Service/Sync.php]: It doesn't look like this is thrown anymore

File: Model/Service/Sync.php [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                $op->setProductIds($ids);
+                $op->delete(); // @codingStandardsIgnoreLine
+                $rowsRemoved = $collection->deleteCurrentItemsByStore($store);
+                $totalDeleted += $rowsRemoved;
+                $this->tickBenchmark(self::BENCHMARK_DELETE_NAME);
+            } catch (\Exception $e) {
+                $this->getLogger()->exception($e);
+            }
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_DELETE_NAME, $store);
```
Comment (supercid) [inline Model/Service/Sync.php]: Try-catch?

File: Util/Benchmark.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    /**
+     * @param string $name
+     * @return float
+     * @throws \Exception
+     */
+    public function getElapsed(string $name)
+    {
+        if (empty($this->times[$name])) {
+            throw new \Exception(sprintf('No such instrumentation: %s', $name));
```
Comment (supercid) [inline Util/Benchmark.php]: How about `Nosto\Exception` or a new benchmark exception?

File: Model/Service/Sync.php [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+     * Handles sync of product collection by sending it to Nosto
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
+     */
+    public function syncIndexedProducts(NostoIndexCollection $collection, Store $store)
+    {
+        if (!$this->nostoDataHelper->isProductUpdatesEnabled($store)) {
```
Comment (olsi-qose) [inline Model/Service/Sync.php]: A return statement should be added in case the condition is met

File: Util/Iterator.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    /**
+     * Iterates through the collection in batches defined by the collection page size
+     * and applies closure to each item
+     * @param \Closure $closure
+     */
+    public function each(\Closure $closure)
+    {
+        $curPage = 1;
```
Comment (olsi-qose) [inline Util/Iterator.php]: Is `$this->collection->setPageSize()` missing? Otherwise it results in one whole page

**Pull request merged**